Rvb.add gameloop

blikkie opened the pull request:
This sets up a basic game loop. When all slots are filled, this will return that the game is a draw.  If a win condition is met before that it set the result to the winner, which will break the game loop. Each turn we prompt the player to drop a stone in a column.

Once a player drops a stone in a column, there is some logic to prevent users from overflowing a column, if there is space, the stone will be placed in the lowest column that's free.

Commit:
Add the skeleton of a game loop

This sets up a basic game loop. When all slots are filled, this will
return that the game is a draw.  If a win condition is met before that
it set the result to the winner, which will break the game loop. Each
turn we prompt the player to drop a stone in a column.

Commit:
Fix array initialization and printing

My initial approach did not create an array with the correct dimentions.
My initial approach of rendering the board actually resulted in
me replacing empty values in the array we're working in with
zeroes, which is bad because I will need these to be empty for future
steps, so instead I'm doing some fancy formatting while printing rows

Commit:
Add logic so stones get dropped in the board

* Switch between player 'x' and player 'o'
* Force players to pick another column if column is full
* When a row is empty the previous row will be written

Commit:
Tell linter to allow ternary expressions

Commit:
Remove log and unused assignment from printBoard

itwasntandy commented on connect-n.js:
```diff
@@ -22,21 +24,87 @@ function createArray(w, h) {
   }
-  array.push(indexRow);
+  for (let i = 0; i < height; i++) {
+    if (!board[i][index]) {
```
Generally its easier to read code, if we can write it like `if (foo) { return } ...` rather than `if (foo) { ... } else { return }``` . In other words if there is a case where you expect to early return, get it out of the way and then continue with the rest of the code.
In this case, I think its clearer if we write it something like this:

```
... 
  for (let i = 0; i < height; i++) {
    if (board[i][index]) {
      board[i - 1][index] = playerSymbol;
      return result;
    }
    result.x = index;
    result.y = I;
  }
  board[height - 1][index] = playerSymbol;
  return result;
}

blikkie replied:
Done

itwasntandy commented on connect-n.js:
```diff
@@ -22,21 +24,87 @@ function createArray(w, h) {
+
+async function gameLoop() {
+  let result;
+  while (!result) {
```
Right now, there might be a slight delay in seeing the result here, as it will only be checked each time the `while` condition is checked - meaning it will need to iterate fully through the inner `for` loop each time.
Perhaps rather than having two nested loops (`while` and `for`), it might be simpler to just use one, with an `if` to check for that condition and `break` when we get a result.

i.e. something like:

```
  let result;
  for (let counter = 0; counter < width * height; counter++) {
    if (result) {
      break;
    }
    ...
  }
```

blikkie replied:
Done

itwasntandy commented:
This looks like a great start. Just left a couple of small suggestions

Commit:
Simplify gameloop and testColumn function

Move return statement up so else clause can be removed
Replace needless while loop with an explicit check
